File: sample_question/custom_validator/sample-validator.rb
EXIT_CODES = {
  "AC" => 0,
  "WA" => 1,
  "TLE" => 2,
  "RE" => 3,
  "PE" => 4,
}.freeze

def finish(exitcode, message = nil)
  if message
    $stderr.puts(message)
    $stderr.flush
  end
  exit(exitcode)
end

def read_line
  line = $stdin.gets
  line&.strip
end

def parse_integer(text)
  return nil if text.nil?

  Integer(text.strip, 10)
rescue ArgumentError
  nil
end

def env_number(name)
  ENV.fetch(name, nil) || "null"
end

def print_limits
  user_language = ENV.fetch("USER_LANGUAGE", nil)
  per_language_limits = ENV.fetch("PER_LANGUAGE_LIMITS", nil)

  $stderr.puts(
    "{" \
    "\"problem_time_limit\": #{env_number('PROBLEM_TIME_LIMIT')}, " \
    "\"problem_output_limit\": #{env_number('PROBLEM_OUTPUT_LIMIT')}, " \
    "\"problem_memory_limit\": #{env_number('PROBLEM_MEMORY_LIMIT')}, " \
    "\"problem_pid_limit\": #{env_number('PROBLEM_PID_LIMIT')}, " \
    "\"user_language\": #{user_language ? "\"#{user_language}\"" : 'null'}, " \
    "\"per_language_limits\": #{per_language_limits || 'null'}" \
    "}"
  )
  $stderr.flush
end

def read_secret_value
  value = parse_integer(read_line)
  finish(EXIT_CODES["RE"], "Invalid secret data") if value.nil?
  value
end

max_value = read_secret_value
max_tries = read_secret_value
secret = read_secret_value

print_limits

$stdout.puts(max_value)
$stdout.flush

iteration = 0
while (user = read_line)
  if user.start_with?("!")
    answer = parse_integer(user[1..])
    finish(EXIT_CODES["RE"], "Invalid data") if answer.nil?
    finish(EXIT_CODES["AC"]) if answer == secret

    finish(EXIT_CODES["WA"], "!#{secret}")
  end

  finish(EXIT_CODES["TLE"], "Exceeded number of allowed iterations") if iteration > max_tries

  value = parse_integer(user)
  finish(EXIT_CODES["RE"], "Invalid data") if value.nil?

  $stdout.puts(secret < value ? "<" : ">=")
  $stdout.flush
  iteration += 1
end

finish(EXIT_CODES["RE"], "Unexpected flow")

File: sample_question/custom_validator/sample-validator.js
"use strict";

const fs = require("fs");

const EXIT_CODES = {
    AC: 0,
    WA: 1,
    TLE: 2,
    RE: 3,
    PE: 4,
};

function finish(exitcode, message) {
    if (message) {
        fs.writeSync(2, message + "\n");
    }
    process.exit(exitcode);
}

// Node has no blocking line reader, so read stdin byte by byte and cut it into lines:
// the conversation is strictly turn-based, and buffering ahead would deadlock it.
let stdinClosed = false;

function readLine() {
    if (stdinClosed) {
        return null;
    }
    const chunk = Buffer.alloc(1);
    let line = "";
    for (;;) {
        let read = 0;
        try {
            read = fs.readSync(0, chunk, 0, 1, null);
        } catch (error) {
            if (error.code === "EAGAIN") {
                continue;
            }
            if (error.code !== "EOF") {
                throw error;
            }
            read = 0;
        }
        if (read === 0) {
            stdinClosed = true;
            return line.length > 0 ? line : null;
        }
        if (chunk[0] === 0x0a) {
            return line;
        }
        line += String.fromCharCode(chunk[0]);
    }
}

function parseInteger(text) {
    const trimmed = text.trim();
    if (!/^[+-]?\d+$/.test(trimmed)) {
        return null;
    }
    return Number(trimmed);
}

function envNumber(name) {
    const value = process.env[name];
    return value === undefined ? "null" : value;
}

function printLimits() {
    const userLanguage = process.env.USER_LANGUAGE;
    const perLanguageLimits = process.env.PER_LANGUAGE_LIMITS;

    const limits =
        "{" +
        `"problem_time_limit": ${envNumber("PROBLEM_TIME_LIMIT")}, ` +
        `"problem_output_limit": ${envNumber("PROBLEM_OUTPUT_LIMIT")}, ` +
        `"problem_memory_limit": ${envNumber("PROBLEM_MEMORY_LIMIT")}, ` +
        `"problem_pid_limit": ${envNumber("PROBLEM_PID_LIMIT")}, ` +
        `"user_language": ${userLanguage === undefined ? "null" : JSON.stringify(userLanguage)}, ` +
        `"per_language_limits": ${perLanguageLimits === undefined ? "null" : perLanguageLimits}` +
        "}";
    fs.writeSync(2, limits + "\n");
}

function readSecretValue() {
    const line = readLine();
    const value = line === null ? null : parseInteger(line);
    if (value === null) {
        finish(EXIT_CODES.RE, "Invalid secret data");
    }
    return value;
}

function main() {
    const maxValue = readSecretValue();
    const maxTries = readSecretValue();
    const secret = readSecretValue();

    printLimits();

    fs.writeSync(1, maxValue + "\n");

    for (let iteration = 0; ; iteration++) {
        const line = readLine();
        if (line === null) {
            break;
        }
        const user = line.trim();

        if (user.startsWith("!")) {
            const answer = parseInteger(user.slice(1));
            if (answer === null) {
                finish(EXIT_CODES.RE, "Invalid data");
            }
            if (answer === secret) {
                finish(EXIT_CODES.AC);
            }
            finish(EXIT_CODES.WA, `!${secret}`);
        }

        if (iteration > maxTries) {
            finish(EXIT_CODES.TLE, "Exceeded number of allowed iterations");
        }
        const value = parseInteger(user);
        if (value === null) {
            finish(EXIT_CODES.RE, "Invalid data");
        }
        fs.writeSync(1, (secret < value ? "<" : ">=") + "\n");
    }

    finish(EXIT_CODES.RE, "Unexpected flow");
}

main();
